controllers/users.js

```diff
@@ -0,0 +1,361 @@
+  var opts = {
+    include_docs: true,
+    key: ['user-settings']
+  };
+  db.medic.view('medic', 'doc_by_type', opts, function(err, results) {
+    if (err) {
+      return callback(err);
+    }
+    callback(null, _.map(results.rows, function(row) {
+      return row.doc;
+    }));
+  });
+};
+
+var getAllUsers = function(callback) {
+  db._users.list({include_docs: true}, function(err, results) {
+    callback(err, results.rows);
+  });
+};
+
+var getFacilities = function(callback) {
+  db.medic.view('medic', 'facilities', {include_docs: true}, function(err, results) {
+    if (err) {
+      return callback(err);
+    }
+    callback(null, _.map(results.rows, function(row) {
+      return row.doc;
+    }));
+  });
+};
+
+var isAPlace = function(place) {
+  return [
+    'district_hospital',
+    'health_center',
+    'clinic'
+  ].indexOf(place.type) !== -1;
+};
+
+var getPlace = function(id, callback) {
+  db.medic.get(id, function(err, place) {
+    if (err) {
+      console.error('Failed to find place.');
+      return callback(err);
+    }
+    if (!isAPlace(place)) {
+      return callback(new Error('Wrong type, this is not a place.'));
+    }
+    callback(null, place);
+  });
+};
+
+var createUser = function(data, response, callback) {
+  response = response || {};
+  var id = createID(data.username),
+      user = getUserUpdates(id, data);
+  db._users.insert(user, id, function(err, body) {
+    if (err) {
+      console.error('Failed to create user.');
+      return callback(err);
+    }
+    response.user = {
+      id: body.id,
+      rev: body.rev
+    };
+    callback(null, data, response);
+  });
+};
+
+/*
+ * Warning: not doing validation of the contact data against a form yet.  The
+ * form is user defined in settings so being liberal with what gets saved to
+ * the database. Ideally CouchDB could validate a given object against a form
+ * in validate_doc_update.
+ */
+var createContact = function(data, response, callback) {
+  response = response || {};
+  db.medic.insert(data.contact, function(err, body) {
+    if (err) {
+      console.error('Failed to create contact.');
+      return callback(err);
+    }
+    // save contact id for user settings
+    data.contact = body.id;
+    response.contact = {
+      id: body.id,
+      rev: body.rev
+    };
+    callback(null, data, response);
+  });
+};
+
+var createUserSettings = function(data, response, callback) {
+  response = response || {};
+  var settings = getSettingsUpdates(data);
+  db.medic.insert(settings, createID(data.username), function(err, body) {
+    if (err) {
+      console.error('Failed to create user settings.');
+      return callback(err);
+    }
+    response['user-settings'] = {
+      id: body.id,
+      rev: body.rev
+    };
+    callback(null, data, response);
+  });
+};
+
+var hasParent = function(facility, id) {
+  // do not modify facility
+  var p = facility.parent;
+  if (facility._id === id) {
+    return true;
+  }
```

Comment:
I think the suggestion here is to do `var p = facility;`, remove the duplicate `id` test, and let the loop body sort it out since `p` would obviously be truthy upon entry into the `while`.